numpy/linalg/linalg.py

```diff
@@ -33,6 +34,28 @@
 from numpy.lib.twodim_base import triu, eye
 from numpy.linalg import _umath_linalg
 
+from numpy._typing import NDArray
+
+class EigResult(NamedTuple):
+    eigenvalues: NDArray
+    eigenvectors: NDArray
+
+class EighResult(NamedTuple):
+    eigenvalues: NDArray
+    eigenvectors: NDArray
+
+class QRResult(NamedTuple):
+    Q: NDArray
```

Comment:
Bit curious about the choice of upper case here and below. Is there a standard for these names?

Reply:
This `NDArray` type name is already present in NumPy. Unless you mean the name of the namedtuple.

Reply:
Public names of type annotations we have, like `np.typing.NDArray`, tend to be upper-case. I think to not confuse them with lower-case objects

Reply:
I was referring to "Q".

Reply:
> Unless you mean the name of the namedtuple.

That was what I was referring to. Looks like NamedTuple is what we want, but am curious if the projects that we want to be compatible with also use it.

Reply:
The uppercase field names are part of the array API spec and were discussed here https://github.com/data-apis/array-api/issues/295. I suppose NumPy could, if it wanted, provide the same fields under multiple names if you really don't like the ones the array API chose, like

```py
class QRResult(NamedTuple):
    Q: NDArray
    R: NDArray

    @property 
    def q(self):
        return self.Q

    @property 
    def r(self):
        return self.R
```

The name of the named tuple is completely arbitrary. We can change it to whatever.

As far as I can tell `NamedTuple` is basically equivalent to the classic `namedtuple`. I think the biggest difference is that NamedTuple supports type annotations, which aren't used in this file yet but presumably they will be at some point. I guess it also makes it easier to add other methods or attributes if you ever wanted to do that. 

I don't think it really makes a difference whether you use NamedTuple or namedtuple. I think it could even be something else like a dataclass, so long as it also has the correct field names and also defines enough methods to act like a tuple.

Reply:
> As far as I can tell NamedTuple is basically equivalent to the classic namedtuple

I think NamedTuple is better, for instance, it allows accessing the fields like so: `a.Q`, but that wouldn't work for `namedtuple`. The declaration of the fields also differs, for instance the declarations here error if `namedtuple` is substituted. The upshot is that I think one or the other should be chosen for the standard for portable code.

Reply:
namedtuple works like that too:

```py
>>> namedtuple('test', ['a', 'b'])(1, 2).a
1
```

I think the only real difference is that NamedTuple lets you specify type annotations. And the class definition is arguably easier to read. As far as the user is concerned though they are identical.